File: R/07_cnv_analysis.R
# 07_cnv_analysis.R
# ==================
# Copy number variation (CNV) inference using inferCNV.
#
# WHAT THIS SCRIPT DOES:
#   1. Load annotated tumour + normal data
#   2. Build the cell annotation file required by inferCNV
#   3. Run inferCNV to infer CNVs from expression patterns
#   4. Classify cells as aneuploid (tumour) or diploid (normal) by CNV score
#   5. Transfer CNV status back to the Seurat object and save
#
# WHEN TO USE:
#   Tumour datasets where you need to distinguish malignant cells from normal
#   stromal/immune cells, or to identify tumour subclones.
#
# IMPORTANT CAVEATS:
#   - inferCNV infers CNV from expression — ~80% concordance with DNA WGS.
#   - Always interpret in context of cell type annotations.
#   - A matched normal sample dramatically improves specificity.
#
# HOW TO USE:
#   1. Edit CONFIGURATION below — especially NORMAL_CELL_TYPES and TUMOUR_CELL_TYPES.
#   2. Run:  Rscript 07_cnv_analysis.R
#
# OUTPUT:
#   results/07_cnv/seurat_cnv.rds
#   results/07_cnv/infercnv_output/   (all inferCNV files)
#   results/07_cnv/plots/

suppressMessages({
  library(Seurat)
  library(infercnv)
  library(ggplot2)
})


# ==============================================================================
#  CONFIGURATION — Edit this section before running
# ==============================================================================

INPUT_PATH    <- "results/03_annotation/seurat_annotated.rds"
CELL_TYPE_COL <- "cell_type"

# --- Reference (normal) and tumour populations --------------------------------
# Normal cells are used as the baseline for CNV inference.
# These should be non-malignant cells (immune, stromal, etc.).
NORMAL_CELL_TYPES <- c("T cell", "B cell", "NK cell")

# Cell types you suspect are tumour / malignant.
TUMOUR_CELL_TYPES <- c("Epithelial")

# --- Gene order file ----------------------------------------------------------
# inferCNV requires a gene ordering file (hg38 or mm10).
# Download from:
#   Human (hg38): https://data.broadinstitute.org/Trinity/CTAT/cnv/hg38_gencode_v27.txt
#   Mouse (mm10): https://data.broadinstitute.org/Trinity/CTAT/cnv/mm10_gencode_vM25.txt
GENE_ORDER_FILE <- "data/raw/hg38_gencode_v27.txt"

# Reference genome. Used for inferCNV annotation.
# Options: "hg38", "hg19", "mm10", "mm9"
GENOME_VERSION <- "hg38"

# --- inferCNV parameters ------------------------------------------------------
# Number of genes per smoothing window.
WINDOW_SIZE <- 101L

# Noise threshold: expression differences smaller than this are treated as noise.
NOISE_FILTER <- 0.1

# CNV score threshold for aneuploid classification.
# Cells with variance of CNV profile > (median_normal + threshold) = Aneuploid.
CNV_SCORE_THRESHOLD <- 0.02

# Number of CPU threads.
N_THREADS <- 4L

# --- Output -------------------------------------------------------------------
OUTPUT_DIR  <- "results/07_cnv"
OUTPUT_PATH <- file.path(OUTPUT_DIR, "seurat_cnv.rds")
INFERCNV_DIR <- file.path(OUTPUT_DIR, "infercnv_output")


# ==============================================================================
#  STEP 1 — LOAD & VALIDATE
# ==============================================================================

load_data <- function() {
  if (!file.exists(INPUT_PATH)) stop("File not found: ", INPUT_PATH)
  seu <- readRDS(INPUT_PATH)

  all_types <- unique(as.character(seu@meta.data[[CELL_TYPE_COL]]))

  missing_normal <- setdiff(NORMAL_CELL_TYPES, all_types)
  missing_tumour <- setdiff(TUMOUR_CELL_TYPES, all_types)

  if (length(missing_normal) > 0L) {
    stop(sprintf("Normal cell types not found: %s\nAvailable: %s",
                 paste(missing_normal, collapse = ", "),
                 paste(all_types, collapse = ", ")))
  }
  if (length(missing_tumour) > 0L) {
    stop(sprintf("Tumour cell types not found: %s\nAvailable: %s",
                 paste(missing_tumour, collapse = ", "),
                 paste(all_types, collapse = ", ")))
  }
  if (!file.exists(GENE_ORDER_FILE)) {
    stop("Gene order file not found: ", GENE_ORDER_FILE,
         "\nDownload from https://data.broadinstitute.org/Trinity/CTAT/cnv/")
  }
  seu
}


# ==============================================================================
#  STEP 2 — BUILD CELL ANNOTATION FILE
# ==============================================================================

build_cell_annotations <- function(seu, out_dir) {
  all_types <- c(NORMAL_CELL_TYPES, TUMOUR_CELL_TYPES)
  mask      <- seu@meta.data[[CELL_TYPE_COL]] %in% all_types
  sub       <- seu[, mask]

  annot_df <- data.frame(
    cell_barcode = colnames(sub),
    cell_type    = as.character(sub@meta.data[[CELL_TYPE_COL]]),
    row.names    = NULL
  )

  annot_path <- file.path(out_dir, "cell_annotations.txt")
  write.table(annot_df, annot_path, sep = "\t",
              quote = FALSE, col.names = FALSE, row.names = FALSE)
  list(sub = sub, annot_path = annot_path)
}


# ==============================================================================
#  STEP 3 — RUN inferCNV
# ==============================================================================

run_infercnv <- function(sub, annot_path) {
  dir.create(INFERCNV_DIR, recursive = TRUE, showWarnings = FALSE)

  # Use raw counts
  counts <- GetAssayData(sub, layer = "counts")

  infercnv_obj <- CreateInfercnvObject(
    raw_counts_matrix    = counts,
    annotations_file     = annot_path,
    delim                = "\t",
    gene_order_file      = GENE_ORDER_FILE,
    ref_group_names      = NORMAL_CELL_TYPES
  )

  infercnv_obj <- run(
    infercnv_obj,
    cutoff               = NOISE_FILTER,
    out_dir              = INFERCNV_DIR,
    cluster_by_groups    = TRUE,
    plot_steps           = FALSE,
    denoise              = TRUE,
    HMM                  = FALSE,
    num_threads          = N_THREADS,
    window_length        = WINDOW_SIZE,
    output_format        = "pdf"
  )

  infercnv_obj
}


# ==============================================================================
#  STEP 4 — CNV SCORE & CLASSIFICATION
# ==============================================================================

classify_cnv <- function(seu, infercnv_obj) {
  # Extract the smoothed expression matrix (CNV signal)
  cnv_mat <- infercnv_obj@expr.data

  # CNV score = variance of CNV profile per cell (higher = more altered genome)
  shared_cells <- intersect(colnames(seu), colnames(cnv_mat))
  cnv_score    <- apply(cnv_mat[, shared_cells, drop = FALSE], 2, var)

  # Reference = normal cell types
  ref_mask     <- seu@meta.data[[CELL_TYPE_COL]] %in% NORMAL_CELL_TYPES
  ref_cells    <- intersect(colnames(seu)[ref_mask], names(cnv_score))
  median_ref   <- median(cnv_score[ref_cells], na.rm = TRUE)

  # Classify
  cnv_status                <- rep(NA_character_, ncol(seu))
  names(cnv_status)         <- colnames(seu)
  cnv_score_full            <- rep(NA_real_,      ncol(seu))
  names(cnv_score_full)     <- colnames(seu)

  cnv_score_full[names(cnv_score)] <- cnv_score
  cnv_status[names(cnv_score)]     <- ifelse(
    cnv_score > median_ref + CNV_SCORE_THRESHOLD,
    "Aneuploid", "Diploid"
  )

  seu$cnv_score  <- cnv_score_full
  seu$cnv_status <- cnv_status

  n_aneupl <- sum(cnv_status == "Aneuploid", na.rm = TRUE)
  n_total  <- sum(!is.na(cnv_status))
  message(sprintf("Aneuploid cells: %d / %d (%.1f%%)",
                  n_aneupl, n_total, 100 * n_aneupl / n_total))
  seu
}


# ==============================================================================
#  STEP 5 — VISUALISATION
# ==============================================================================

save_cnv_plots <- function(seu, plot_dir) {
  dir.create(plot_dir, recursive = TRUE, showWarnings = FALSE)

  if ("umap" %in% names(seu@reductions) && "cnv_status" %in% colnames(seu@meta.data)) {
    p_status <- DimPlot(
      seu, reduction = "umap", group.by = "cnv_status", pt.size = 0.4,
      cols = c("Aneuploid" = "#C44E52", "Diploid" = "#4C72B0")
    ) + ggtitle("CNV Status on UMAP") + theme_classic()
    ggsave(file.path(plot_dir, "01_umap_cnv_status.pdf"), p_status, width = 8, height = 7)

    p_score <- FeaturePlot(seu, features = "cnv_score", reduction = "umap",
                           pt.size = 0.4) +
      scale_colour_gradient(low = "grey90", high = "#C44E52") +
      ggtitle("CNV Score on UMAP") + theme_classic()
    ggsave(file.path(plot_dir, "02_umap_cnv_score.pdf"), p_score, width = 8, height = 7)
  }
}


# ==============================================================================
#  MAIN
# ==============================================================================

main <- function() {
  dir.create(OUTPUT_DIR, recursive = TRUE, showWarnings = FALSE)
  plot_dir <- file.path(OUTPUT_DIR, "plots")

  seu                    <- load_data()
  result                 <- build_cell_annotations(seu, OUTPUT_DIR)
  infercnv_obj           <- run_infercnv(result$sub, result$annot_path)
  seu                    <- classify_cnv(seu, infercnv_obj)
  save_cnv_plots(seu, plot_dir)

  saveRDS(seu, OUTPUT_PATH)
  message("Saved -> ", OUTPUT_PATH)
}

main()

File: Python/07_cnv_analysis.py
#!/usr/bin/env python3
"""
07_cnv_analysis.py
==================
Copy number variation (CNV) analysis using inferCNVpy.

WHAT THIS SCRIPT DOES:
  1. Load annotated tumour + normal data
  2. Annotate genes with chromosome positions (via Ensembl or a custom file)
  3. Run inferCNVpy to infer CNV from expression patterns
  4. Classify cells as aneuploid (tumour) vs. diploid (normal)
  5. Save CNV-annotated AnnData

WHEN TO USE:
  Tumour datasets where you need to distinguish malignant cells from normal
  stromal/immune cells, or to identify tumour subclones.

IMPORTANT CAVEATS:
  - inferCNVpy infers CNV from expression — ~80% concordance with DNA WGS
  - Always interpret in context of cell type annotations
  - A matched normal sample dramatically improves specificity

HOW TO USE:
  1. Edit CONFIGURATION below — especially NORMAL_CELL_TYPES and TUMOUR_CELL_TYPES.
  2. Run:  python 07_cnv_analysis.py

OUTPUT:
  results/07_cnv/adata_cnv.h5ad
  results/07_cnv/plots/
"""

import sys
import warnings
from pathlib import Path

import numpy as np
import pandas as pd
import scanpy as sc
import plotly.express as px

sys.path.insert(0, str(Path("..") / "src"))
from sc_workflow.utils import setup_scanpy, save_adata

warnings.filterwarnings("ignore")
setup_scanpy()


# ==============================================================================
#  CONFIGURATION — Edit this section before running
# ==============================================================================

INPUT_PATH = "results/03_annotation/adata_annotated.h5ad"
CELL_TYPE_COL = "cell_type"

# --- Reference (normal) and tumour populations --------------------------------
# Normal cells are used as the baseline for CNV inference.
# These should be non-malignant cells (immune, stromal, etc.).
NORMAL_CELL_TYPES = [
    "T cell",
    "B cell",
    "NK cell",
]

# Cell types you suspect are tumour / malignant.
TUMOUR_CELL_TYPES = [
    "Epithelial",
]

# --- Gene position annotation -------------------------------------------------
# Leave empty ("") to auto-download positions from Ensembl via pybiomart.
# Or provide a path to a custom TSV file with columns:
#   gene_name  chromosome  start  end    (no header, tab-separated)
GENE_ORDER_FILE = ""

# Reference genome version for Ensembl lookup.
# Options: "GRCh38", "GRCh37", "GRCm39", "GRCm38"
GENOME_VERSION = "GRCh38"

# --- inferCNVpy parameters ----------------------------------------------------
# Number of genes per sliding window (smoothing bandwidth).
WINDOW_SIZE = 100

# Step size between windows.
STEP_SIZE = 10

# CNV score threshold for classifying a cell as aneuploid.
# Cells with CNV score > (median_normal_score + threshold) are called aneuploid.
CNV_SCORE_THRESHOLD = 0.1

# --- Output ------------------------------------------------------------------
OUTPUT_PATH = "results/07_cnv/adata_cnv.h5ad"


# ==============================================================================
#  STEP 1 — LOAD DATA
# ==============================================================================

def load_data():
    p = Path(INPUT_PATH)
    if not p.exists():
        print(f"File not found: {p}")
        sys.exit(1)
    adata = sc.read_h5ad(p)

    all_types = adata.obs[CELL_TYPE_COL].unique().tolist()
    missing_normal = [t for t in NORMAL_CELL_TYPES if t not in all_types]
    missing_tumour = [t for t in TUMOUR_CELL_TYPES if t not in all_types]

    if missing_normal:
        print(f"Normal types not found: {missing_normal}\nAvailable: {all_types}")
        sys.exit(1)
    if missing_tumour:
        print(f"Tumour types not found: {missing_tumour}\nAvailable: {all_types}")
        sys.exit(1)

    return adata


# ==============================================================================
#  STEP 2 — GENE CHROMOSOME POSITIONS
# ==============================================================================

def annotate_gene_positions(adata):
    if GENE_ORDER_FILE:
        gf = Path(GENE_ORDER_FILE)
        if not gf.exists():
            print(f"Gene order file not found: {gf}")
            sys.exit(1)
        gene_pos = pd.read_csv(gf, sep="\t", header=None,
                               names=["gene_name", "chromosome", "start", "end"],
                               index_col=0)
    else:
        try:
            import pybiomart as pbm
            dataset_name = {
                "GRCh38": "hsapiens_gene_ensembl",
                "GRCh37": "hsapiens_gene_ensembl",
                "GRCm39": "mmusculus_gene_ensembl",
                "GRCm38": "mmusculus_gene_ensembl",
            }[GENOME_VERSION]
            server = pbm.Server(host="http://www.ensembl.org")
            mart = server["ENSEMBL_MART_ENSEMBL"][dataset_name]
            gene_pos = mart.query(
                attributes=["hgnc_symbol", "chromosome_name", "start_position", "end_position"]
            ).dropna()
            gene_pos.columns = ["gene_name", "chromosome", "start", "end"]
            gene_pos = (
                gene_pos[gene_pos["gene_name"].isin(adata.var_names)]
                .drop_duplicates("gene_name")
                .set_index("gene_name")
            )
        except ImportError:
            print("pybiomart not installed (pip install pybiomart). "
                  "Provide a custom GENE_ORDER_FILE instead.")
            sys.exit(1)
        except Exception as e:
            print(f"Ensembl download failed: {e}\n"
                  "Check internet connection, or provide a GENE_ORDER_FILE.")
            sys.exit(1)

    for col in ["chromosome", "start", "end"]:
        if col in gene_pos.columns:
            adata.var[col] = gene_pos.reindex(adata.var_names)[col]

    return adata


# ==============================================================================
#  STEP 3 — RUN inferCNVpy
# ==============================================================================

def run_infercnv(adata):
    try:
        import infercnvpy as cnv
    except ImportError:
        print("infercnvpy not installed (pip install infercnvpy).")
        sys.exit(1)

    all_interest = NORMAL_CELL_TYPES + TUMOUR_CELL_TYPES
    mask = adata.obs[CELL_TYPE_COL].isin(all_interest)
    adata_cnv = adata[mask].copy()

    REF_KEY = "is_reference"
    adata_cnv.obs[REF_KEY] = adata_cnv.obs[CELL_TYPE_COL].isin(NORMAL_CELL_TYPES)

    if "counts" in adata_cnv.layers:
        import scipy.sparse as sp
        adata_cnv.X = adata_cnv.layers["counts"]

    cnv.tl.infercnv(
        adata_cnv,
        reference_key=REF_KEY,
        reference_cat=True,
        window_size=WINDOW_SIZE,
        step=STEP_SIZE,
        key_added="cnv",
    )
    cnv.tl.pca(adata_cnv, use_rep="X_cnv")
    cnv.tl.umap(adata_cnv)
    cnv.tl.leiden(adata_cnv)

    cnv_mat = adata_cnv.obsm["X_cnv"]
    adata_cnv.obs["cnv_score"] = np.var(
        cnv_mat.toarray() if hasattr(cnv_mat, "toarray") else cnv_mat,
        axis=1,
    )

    med_ref = adata_cnv.obs.loc[adata_cnv.obs[REF_KEY], "cnv_score"].median()
    adata_cnv.obs["cnv_status"] = np.where(
        adata_cnv.obs["cnv_score"] > med_ref + CNV_SCORE_THRESHOLD,
        "Aneuploid",
        "Diploid",
    )

    adata.obs.loc[adata_cnv.obs.index, "cnv_score"] = adata_cnv.obs["cnv_score"]
    adata.obs.loc[adata_cnv.obs.index, "cnv_status"] = adata_cnv.obs["cnv_status"]

    n_aneupl = (adata_cnv.obs["cnv_status"] == "Aneuploid").sum()
    pct = 100 * n_aneupl / len(adata_cnv.obs)
    print(f"Aneuploid cells: {n_aneupl:,} ({pct:.1f}%)")
    return adata, adata_cnv


# ==============================================================================
#  STEP 4 — VISUALISATION
# ==============================================================================

def save_cnv_plots(adata, adata_cnv, plot_dir):
    plot_dir = Path(plot_dir)
    plot_dir.mkdir(parents=True, exist_ok=True)

    if "X_umap" in adata.obsm and "cnv_status" in adata.obs.columns:
        umap = adata.obsm["X_umap"]
        obs = adata.obs.copy()
        obs["UMAP1"] = umap[:, 0]
        obs["UMAP2"] = umap[:, 1]
        fig = px.scatter(
            obs, x="UMAP1", y="UMAP2", color="cnv_status",
            color_discrete_map={"Aneuploid": "#C44E52", "Diploid": "#4C72B0"},
            title="CNV Status on Original UMAP",
            template="plotly_white", opacity=0.7,
        )
        fig.update_traces(marker_size=3)
        fig.update_layout(height=480)
        fig.write_html(plot_dir / "01_umap_cnv_status.html")

    try:
        import infercnvpy as cnv
        import matplotlib.pyplot as plt

        fig, axs = plt.subplots(1, 2, figsize=(14, 5))
        cnv.pl.umap(adata_cnv, color="cnv_status", ax=axs[0], show=False,
                    title="CNV Status")
        cnv.pl.umap(adata_cnv, color="cnv_score", ax=axs[1], show=False,
                    title="CNV Score", color_map="Reds")
        plt.tight_layout()
        fig.savefig(plot_dir / "02_cnv_umap.png", bbox_inches="tight", dpi=110)
        plt.close(fig)

        fig2 = cnv.pl.chromosome_heatmap(adata_cnv, groupby="cnv_status",
                                          return_fig=True)
        fig2.savefig(plot_dir / "03_chromosome_heatmap.png", bbox_inches="tight", dpi=100)
        plt.close(fig2)
    except Exception as e:
        print(f"CNV plot generation failed: {e}")


# ==============================================================================
#  MAIN
# ==============================================================================

def main():
    plot_dir = Path(OUTPUT_PATH).parent / "plots"

    adata = load_data()
    adata = annotate_gene_positions(adata)
    adata, adata_cnv = run_infercnv(adata)
    save_cnv_plots(adata, adata_cnv, plot_dir)

    save_adata(adata_cnv, OUTPUT_PATH)
    print(f"Saved → {OUTPUT_PATH}")


if __name__ == "__main__":
    main()
